<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8"/><title>Assignment 3 - Hidden Markov Models</title><style>
/* webkit printing magic: print all background colors */
html {
	-webkit-print-color-adjust: exact;
}
* {
	box-sizing: border-box;
	-webkit-print-color-adjust: exact;
}

html,
body {
	margin: 0;
	padding: 0;
}
@media only screen {
	body {
		margin: 2em auto;
		max-width: 900px;
		color: rgb(55, 53, 47);
	}
}

body {
	line-height: 1.5;
	white-space: pre-wrap;
}

a,
a.visited {
	color: inherit;
	text-decoration: underline;
}

.pdf-relative-link-path {
	font-size: 80%;
	color: #444;
}

h1,
h2,
h3 {
	letter-spacing: -0.01em;
	line-height: 1.2;
	font-weight: 600;
	margin-bottom: 0;
}

.page-title {
	font-size: 2.5rem;
	font-weight: 700;
	margin-top: 0;
	margin-bottom: 0.75em;
}

h1 {
	font-size: 1.875rem;
	margin-top: 1.875rem;
}

h2 {
	font-size: 1.5rem;
	margin-top: 1.5rem;
}

h3 {
	font-size: 1.25rem;
	margin-top: 1.25rem;
}

.source {
	border: 1px solid #ddd;
	border-radius: 3px;
	padding: 1.5em;
	word-break: break-all;
}

.callout {
	border-radius: 3px;
	padding: 1rem;
}

figure {
	margin: 1.25em 0;
	page-break-inside: avoid;
}

figcaption {
	opacity: 0.5;
	font-size: 85%;
	margin-top: 0.5em;
}

mark {
	background-color: transparent;
}

.indented {
	padding-left: 1.5em;
}

hr {
	background: transparent;
	display: block;
	width: 100%;
	height: 1px;
	visibility: visible;
	border: none;
	border-bottom: 1px solid rgba(55, 53, 47, 0.09);
}

img {
	max-width: 100%;
}

@media only print {
	img {
		max-height: 100vh;
		object-fit: contain;
	}
}

@page {
	margin: 1in;
}

.collection-content {
	font-size: 0.875rem;
}

.column-list {
	display: flex;
	justify-content: space-between;
}

.column {
	padding: 0 1em;
}

.column:first-child {
	padding-left: 0;
}

.column:last-child {
	padding-right: 0;
}

.table_of_contents-item {
	display: block;
	font-size: 0.875rem;
	line-height: 1.3;
	padding: 0.125rem;
}

.table_of_contents-indent-1 {
	margin-left: 1.5rem;
}

.table_of_contents-indent-2 {
	margin-left: 3rem;
}

.table_of_contents-indent-3 {
	margin-left: 4.5rem;
}

.table_of_contents-link {
	text-decoration: none;
	opacity: 0.7;
	border-bottom: 1px solid rgba(55, 53, 47, 0.18);
}

table,
th,
td {
	border: 1px solid rgba(55, 53, 47, 0.09);
	border-collapse: collapse;
}

table {
	border-left: none;
	border-right: none;
}

th,
td {
	font-weight: normal;
	padding: 0.25em 0.5em;
	line-height: 1.5;
	min-height: 1.5em;
	text-align: left;
}

th {
	color: rgba(55, 53, 47, 0.6);
}

ol,
ul {
	margin: 0;
	margin-block-start: 0.6em;
	margin-block-end: 0.6em;
}

li > ol:first-child,
li > ul:first-child {
	margin-block-start: 0.6em;
}

ul > li {
	list-style: disc;
}

ul.to-do-list {
	text-indent: -1.7em;
}

ul.to-do-list > li {
	list-style: none;
}

.to-do-children-checked {
	text-decoration: line-through;
	opacity: 0.375;
}

ul.toggle > li {
	list-style: none;
}

ul {
	padding-inline-start: 1.7em;
}

ul > li {
	padding-left: 0.1em;
}

ol {
	padding-inline-start: 1.6em;
}

ol > li {
	padding-left: 0.2em;
}

.mono ol {
	padding-inline-start: 2em;
}

.mono ol > li {
	text-indent: -0.4em;
}

.toggle {
	padding-inline-start: 0em;
	list-style-type: none;
}

/* Indent toggle children */
.toggle > li > details {
	padding-left: 1.7em;
}

.toggle > li > details > summary {
	margin-left: -1.1em;
}

.selected-value {
	display: inline-block;
	padding: 0 0.5em;
	background: rgba(206, 205, 202, 0.5);
	border-radius: 3px;
	margin-right: 0.5em;
	margin-top: 0.3em;
	margin-bottom: 0.3em;
	white-space: nowrap;
}

.collection-title {
	display: inline-block;
	margin-right: 1em;
}

time {
	opacity: 0.5;
}

.icon {
	display: inline-block;
	max-width: 1.2em;
	max-height: 1.2em;
	text-decoration: none;
	vertical-align: text-bottom;
	margin-right: 0.5em;
}

img.icon {
	border-radius: 3px;
}

.user-icon {
	width: 1.5em;
	height: 1.5em;
	border-radius: 100%;
	margin-right: 0.5rem;
}

.user-icon-inner {
	font-size: 0.8em;
}

.text-icon {
	border: 1px solid #000;
	text-align: center;
}

.page-cover-image {
	display: block;
	object-fit: cover;
	width: 100%;
	height: 30vh;
}

.page-header-icon {
	font-size: 3rem;
	margin-bottom: 1rem;
}

.page-header-icon-with-cover {
	margin-top: -0.72em;
	margin-left: 0.07em;
}

.page-header-icon img {
	border-radius: 3px;
}

.link-to-page {
	margin: 1em 0;
	padding: 0;
	border: none;
	font-weight: 500;
}

p > .user {
	opacity: 0.5;
}

td > .user,
td > time {
	white-space: nowrap;
}

input[type="checkbox"] {
	transform: scale(1.5);
	margin-right: 0.6em;
	vertical-align: middle;
}

p {
	margin-top: 0.5em;
	margin-bottom: 0.5em;
}

.image {
	border: none;
	margin: 1.5em 0;
	padding: 0;
	border-radius: 0;
	text-align: center;
}

.code,
code {
	background: rgba(135, 131, 120, 0.15);
	border-radius: 3px;
	padding: 0.2em 0.4em;
	border-radius: 3px;
	font-size: 85%;
	tab-size: 2;
}

code {
	color: #eb5757;
}

.code {
	padding: 1.5em 1em;
}

.code > code {
	background: none;
	padding: 0;
	font-size: 100%;
	color: inherit;
}

blockquote {
	font-size: 1.25em;
	margin: 1em 0;
	padding-left: 1em;
	border-left: 3px solid rgb(55, 53, 47);
}

.bookmark-href {
	font-size: 0.75em;
	opacity: 0.5;
}

.sans { font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Helvetica, "Apple Color Emoji", Arial, sans-serif, "Segoe UI Emoji", "Segoe UI Symbol"; }
.code { font-family: 'SFMono-Regular', Consolas, 'Liberation Mono', Menlo, Courier, monospace; }
.serif { font-family: Lyon-Text, Georgia, KaiTi, STKaiTi, '华文楷体', KaiTi_GB2312, '楷体_GB2312', serif; }
.mono { font-family: Nitti, 'Microsoft YaHei', '微软雅黑', monospace; }
.pdf .sans { font-family: Inter, -apple-system, BlinkMacSystemFont, "Segoe UI", Helvetica, "Apple Color Emoji", Arial, sans-serif, "Segoe UI Emoji", "Segoe UI Symbol", 'Twemoji', 'Noto Color Emoji', 'Noto Sans CJK SC', 'Noto Sans CJK KR'; }

.pdf .code { font-family: Source Code Pro, 'SFMono-Regular', Consolas, 'Liberation Mono', Menlo, Courier, monospace, 'Twemoji', 'Noto Color Emoji', 'Noto Sans Mono CJK SC', 'Noto Sans Mono CJK KR'; }

.pdf .serif { font-family: PT Serif, Lyon-Text, Georgia, KaiTi, STKaiTi, '华文楷体', KaiTi_GB2312, '楷体_GB2312', serif, 'Twemoji', 'Noto Color Emoji', 'Noto Sans CJK SC', 'Noto Sans CJK KR'; }

.pdf .mono { font-family: PT Mono, Nitti, 'Microsoft YaHei', '微软雅黑', monospace, 'Twemoji', 'Noto Color Emoji', 'Noto Sans Mono CJK SC', 'Noto Sans Mono CJK KR'; }

.highlight-default {
}
.highlight-gray {
	color: rgb(155,154,151);
}
.highlight-brown {
	color: rgb(100,71,58);
}
.highlight-orange {
	color: rgb(217,115,13);
}
.highlight-yellow {
	color: rgb(223,171,1);
}
.highlight-teal {
	color: rgb(15,123,108);
}
.highlight-blue {
	color: rgb(11,110,153);
}
.highlight-purple {
	color: rgb(105,64,165);
}
.highlight-pink {
	color: rgb(173,26,114);
}
.highlight-red {
	color: rgb(224,62,62);
}
.highlight-gray_background {
	background: rgb(235,236,237);
}
.highlight-brown_background {
	background: rgb(233,229,227);
}
.highlight-orange_background {
	background: rgb(250,235,221);
}
.highlight-yellow_background {
	background: rgb(251,243,219);
}
.highlight-teal_background {
	background: rgb(221,237,234);
}
.highlight-blue_background {
	background: rgb(221,235,241);
}
.highlight-purple_background {
	background: rgb(234,228,242);
}
.highlight-pink_background {
	background: rgb(244,223,235);
}
.highlight-red_background {
	background: rgb(251,228,228);
}
.block-color-default {
	color: inherit;
	fill: inherit;
}
.block-color-gray {
	color: rgba(55, 53, 47, 0.6);
	fill: rgba(55, 53, 47, 0.6);
}
.block-color-brown {
	color: rgb(100,71,58);
	fill: rgb(100,71,58);
}
.block-color-orange {
	color: rgb(217,115,13);
	fill: rgb(217,115,13);
}
.block-color-yellow {
	color: rgb(223,171,1);
	fill: rgb(223,171,1);
}
.block-color-teal {
	color: rgb(15,123,108);
	fill: rgb(15,123,108);
}
.block-color-blue {
	color: rgb(11,110,153);
	fill: rgb(11,110,153);
}
.block-color-purple {
	color: rgb(105,64,165);
	fill: rgb(105,64,165);
}
.block-color-pink {
	color: rgb(173,26,114);
	fill: rgb(173,26,114);
}
.block-color-red {
	color: rgb(224,62,62);
	fill: rgb(224,62,62);
}
.block-color-gray_background {
	background: rgb(235,236,237);
}
.block-color-brown_background {
	background: rgb(233,229,227);
}
.block-color-orange_background {
	background: rgb(250,235,221);
}
.block-color-yellow_background {
	background: rgb(251,243,219);
}
.block-color-teal_background {
	background: rgb(221,237,234);
}
.block-color-blue_background {
	background: rgb(221,235,241);
}
.block-color-purple_background {
	background: rgb(234,228,242);
}
.block-color-pink_background {
	background: rgb(244,223,235);
}
.block-color-red_background {
	background: rgb(251,228,228);
}

.checkbox {
	display: inline-flex;
	vertical-align: text-bottom;
	width: 16;
	height: 16;
	background-size: 16px;
	margin-left: 2px;
	margin-right: 5px;
}

.checkbox-on {
	background-image: url("data:image/svg+xml;charset=UTF-8,%3Csvg%20width%3D%2216%22%20height%3D%2216%22%20viewBox%3D%220%200%2016%2016%22%20fill%3D%22none%22%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%3E%0A%3Crect%20width%3D%2216%22%20height%3D%2216%22%20fill%3D%22%2358A9D7%22%2F%3E%0A%3Cpath%20d%3D%22M6.71429%2012.2852L14%204.9995L12.7143%203.71436L6.71429%209.71378L3.28571%206.2831L2%207.57092L6.71429%2012.2852Z%22%20fill%3D%22white%22%2F%3E%0A%3C%2Fsvg%3E");
}

.checkbox-off {
	background-image: url("data:image/svg+xml;charset=UTF-8,%3Csvg%20width%3D%2216%22%20height%3D%2216%22%20viewBox%3D%220%200%2016%2016%22%20fill%3D%22none%22%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%3E%0A%3Crect%20x%3D%220.75%22%20y%3D%220.75%22%20width%3D%2214.5%22%20height%3D%2214.5%22%20fill%3D%22white%22%20stroke%3D%22%2336352F%22%20stroke-width%3D%221.5%22%2F%3E%0A%3C%2Fsvg%3E");
}
	
</style></head><body><article id="6ba9dfa4-8465-454c-9705-bd15cf0b3ec5" class="page serif"><header><h1 class="page-title">Assignment 3 - Hidden Markov Models</h1></header><div class="page-body"><p id="1bea38e6-b3b1-45c7-9254-31e4e9bf63fb" class=""><strong>COMP 561 - Computational Biology Methods and Research</strong></p><p id="50b27785-73e1-4dac-b784-4db65be7da99" class=""><strong>LE, Nhat Hung
</strong>McGill ID: 260793376
Date: November 4, 2019
Due date: November 11, 2019</p><p id="2247b6d8-d39b-47ff-9f71-f0dc1c0a9a10" class="">Prof. Mathieu Blanchette
Fall 2019</p><hr id="561324c1-6767-4a72-bb29-1fd196625192"/><h2 id="950508e0-f8b6-4252-b9f5-884c3ec18af5" class="">Question 1 (80 points).</h2><p id="7f299eb5-8069-4f1a-95de-65ab11e76a86" class="">In this question, you will implement the Viterbi algorithm and apply it to the gene finding HMM seen in class to predict genes in a bacterial genome.</p><p id="d68c4c94-24d5-47e1-ab06-8ed7d362ed65" class=""><em>Vibrio cholerae </em>is the bacteria that causes cholera. It was first sequenced in 2000 ( see https://www.nature.com/articles/35020000 ). Download the genome sequence here:</p><p id="d060f620-39ff-4e17-9ac5-dcd94bee8842" class="">ftp://ftp.ensemblgenomes.org/pub/bacteria/release-45/fasta/bacteria_81_collection/vibrio_cholerae/dna/Vibrio_cholerae.GFC_11.dna.toplevel.fa.gz</p><p id="8372222e-f6dc-42fd-bfa0-5d46e540f86c" class="">Note: Although the genome of <em>Vibrio cholera </em>consists of two chromosomes, the sequence file contains more than two sequences, because it is incompletely assembled.</p><p id="ee40a5df-1021-4df4-88f1-69a76c355a44" class="">Download the gene annotation here:</p><p id="9c108031-80c2-4f3e-8649-d9664612b867" class="">ftp://ftp.ensemblgenomes.org/pub/bacteria/release-45/gff3/bacteria_81_collection/vibrio_cholerae/Vibrio_cholerae.GFC_11.37.gff3.gz</p><p id="4dd2bdc8-1a96-4d73-8c21-a92ea7620797" class="">The format of both files should be fairly self-explanatory, but you can find more information about the gff3 file format here: http://gmod.org/wiki/GFF3. One important detail we’ve omitted in our in-class discussion is that genes can actually be located either on the forward or the reverse DNA strand. In class, we have only considered the genes located on the forward strand. In this assignment, we will also only focus on the genes on the forward strand. It is not much more complicated to handle genes on the negative strand, but it’s just annoying, so we will pretend that genes on the negative strand do not
exist, i.e. that portions of the genome occupied by genes on the reverse strand are actually intergenic regions.</p><p id="f53b8cd9-dd64-4210-a4ae-e51c483f7e3b" class="">a) (10 points) Using the genome sequence and gene annotation for Vibrio cholerae, infer the (i) average length of intergenic regions, (ii) average length of genic region, (iii) the nucleotide frequency table for intergenic regions; (iv) the codon frequency table for genic regions.</p><h3 id="0e224442-77b4-4e92-b96b-c9788757a498" class="">Solution:</h3><p id="1e9e2bc6-93e7-416e-97c4-c7ff29f7d85c" class="">(i)   Average length of intergenic regions: 1290</p><p id="09c439fd-2183-41ad-9b88-d6c85a68e737" class="">(ii)  Average length of genic region: 991</p><p id="da3e7222-956a-48a2-9813-8b30d27a12e5" class="">(iii) Nucleotide frequency table for intergenic regions:<div class="indented"><pre id="6d47e662-de98-44da-bf34-db2d316d1231" class="code"><code>{&#x27;A&#x27;: 0.266, &#x27;C&#x27;: 0.243, &#x27;G&#x27;: 0.226, &#x27;T&#x27;: 0.265}</code></pre></div></p><p id="365dcb13-2f35-4405-b9e8-ffe695d71840" class="">
</p><p id="8a84f69d-e2e7-441f-b22f-324245302810" class="">(iv) Codon frequency table for genic regions:<div class="indented"><pre id="5d2a238a-385f-4332-b88c-64566d12a6a5" class="code"><code>{&#x27;AAA&#x27;: 0.036, &#x27;AAC&#x27;: 0.02, &#x27;AAG&#x27;: 0.014, &#x27;AAT&#x27;: 0.019, &#x27;ACA&#x27;: 0.008, &#x27;ACC&#x27;: 0.021,
 &#x27;ACG&#x27;: 0.011, &#x27;ACT&#x27;: 0.013, &#x27;AGA&#x27;: 0.003, &#x27;AGC&#x27;: 0.014, &#x27;AGG&#x27;: 0.001, &#x27;AGT&#x27;: 0.011,
 &#x27;ATA&#x27;: 0.004, &#x27;ATC&#x27;: 0.025, &#x27;ATG&#x27;: 0.026, &#x27;ATT&#x27;: 0.031, &#x27;CAA&#x27;: 0.033, &#x27;CAC&#x27;: 0.011,
 &#x27;CAG&#x27;: 0.018, &#x27;CAT&#x27;: 0.013, &#x27;CCA&#x27;: 0.013, &#x27;CCC&#x27;: 0.006, &#x27;CCG&#x27;: 0.011, &#x27;CCT&#x27;: 0.011,
 &#x27;CGA&#x27;: 0.005, &#x27;CGC&#x27;: 0.018, &#x27;CGG&#x27;: 0.003, &#x27;CGT&#x27;: 0.02, &#x27;CTA&#x27;: 0.009, &#x27;CTC&#x27;: 0.014,
 &#x27;CTG&#x27;: 0.029, &#x27;CTT&#x27;: 0.013, &#x27;GAA&#x27;: 0.039, &#x27;GAC&#x27;: 0.014, &#x27;GAG&#x27;: 0.024, &#x27;GAT&#x27;: 0.037,
 &#x27;GCA&#x27;: 0.019, &#x27;GCC&#x27;: 0.022, &#x27;GCG&#x27;: 0.03, &#x27;GCT&#x27;: 0.021, &#x27;GGA&#x27;: 0.007, &#x27;GGC&#x27;: 0.025,
 &#x27;GGG&#x27;: 0.009, &#x27;GGT&#x27;: 0.027, &#x27;GTA&#x27;: 0.011, &#x27;GTC&#x27;: 0.014, &#x27;GTG&#x27;: 0.029, &#x27;GTT&#x27;: 0.016,
 &#x27;TAA&#x27;: 0.002, &#x27;TAC&#x27;: 0.014, &#x27;TAG&#x27;: 0.001, &#x27;TAT&#x27;: 0.016, &#x27;TCA&#x27;: 0.01, &#x27;TCC&#x27;: 0.006,
 &#x27;TCG&#x27;: 0.009, &#x27;TCT&#x27;: 0.011, &#x27;TGA&#x27;: 0.001, &#x27;TGC&#x27;: 0.004, &#x27;TGG&#x27;: 0.013, &#x27;TGT&#x27;: 0.006,
 &#x27;TTA&#x27;: 0.019, &#x27;TTC&#x27;: 0.014, &#x27;TTG&#x27;: 0.023, &#x27;TTT&#x27;: 0.026}</code></pre></div></p><p id="eed6e41a-7561-4fd0-95b7-dc8e59a75d92" class="">
</p><p id="55881930-09e1-4ecc-82f3-460ec5d2fa18" class="">b) (30 points) Using the programming language of your choice, implement the Viterbi algorithm. Your program does not need to work with an arbitrary HMM, but only needs to work for the specific bacterial gene-finding HMM seen in class (4 states: Intergenic, Start, Middle, Stop). The program’s argument should be</p><p id="d3b3a59a-0d1e-43fa-8baf-fbe42059d6d4" class="">I. A Fasta file containing one or more bacterial genome sequences.</p><p id="38c06bab-e371-4e29-9557-8fe70c490b1c" class="">II. A configuration file (formatted as you want), which contains the information found in (a), to be used to determine the emission and transition probabilities for your HMM. (Assume that the initial state probability is 1 for the intergenic state, and zero for all other states). The output should be a GFF3 file with one line per predicted gene (only lines corresponding to portions annotated as genes need to be included in the output file), e.g. something like:
</p><pre id="542ab400-ca92-4b21-a5d6-658f5e3a24e8" class="code"><code>DN38.contig00011 ena CDS 85269 86705 . + 0 .
DN38.contig00011 ena CDS 87006 87230 . + 0 .
DN38.contig00011 ena CDS 88079 89323 . + 0 .</code></pre><p id="bda31e6e-5fd9-4694-86e1-4c18e3cc2938" class="">Submit your code, along with a simple explanation about how to run it.</p><h3 id="a70533c5-a34d-4efd-a51c-b1b75ed6f889" class="">Solution:</h3><p id="4b0c50e0-7f17-44ad-be14-9a45ee529e7c" class="">Refer to the code and instructions in <code>README.md</code>.</p><p id="7940cf62-a748-4b6b-b516-27d3d1394058" class="">
</p><p id="7b90a0b2-b7e2-4b9f-84af-53edf968e6ef" class="">c) (10 points) <em>Vibrio vulnificus </em>is a species of bacteria closely related to <em>Vibrio cholerae</em>. In 2005, it caused an outbreak following the hurricane Katrina (https://www.ncbi.nlm.nih.gov/pubmed/16195696 ). Its genome was sequenced and can be downloaded here:</p><p id="7548d02e-dc6d-40ea-9b8d-5b4b70c60b18" class="">ftp://ftp.ensemblgenomes.org/pub/bacteria/release-45/fasta/bacteria_64_collection/vibrio_vulnificus/dna/Vibrio_vulnificus.ASM74310v1.dna.toplevel.fa.gz</p><p id="696c0498-54db-47f7-9c55-c80a3082e154" class="">Run your program on this genome, using the parameters obtained for <em>Vibrio cholerae</em>. Submit the GFF3 file with your gene predictions (forward strand only).</p><h3 id="1171d466-7abb-4487-a1ab-903b85a57051" class="">Solution:</h3><p id="37d11ff1-1016-4b98-bd46-afea5223a68a" class="">Reproduce the output by following the instructions in <code>README.md</code>.</p><p id="7693d93c-3a25-4c6a-bbf4-896f85105fdc" class="">
</p><p id="156b9926-9731-417b-b397-62f77dd67cb6" class="">d) (15 points) The genes of <em>Vibrio vulnificus </em>were annotated based on a variety of types of evidence. You can download the GFF3 file here:</p><p id="761c5086-13e7-4145-b38a-c583b0a43d6f" class="">ftp://ftp.ensemblgenomes.org/pub/bacteria/release-45/gff3/bacteria_64_collection/vibrio_vulnificus/Vibrio_vulnificus.ASM74310v1.37.gff3.gz</p><p id="0b90bcbb-475d-4035-9a40-91b314329432" class="">Evaluate the accuracy of your predictions against the gene annotation available here. Specifically, report:</p><ul id="4da19efb-d956-4442-80a5-68c0569706bb" class="bulleted-list"><li>The fraction of annotated genes that:<ul id="7e3ff989-88c3-4b8e-b463-2928fddbee2d" class="bulleted-list"><li>Perfectly match both ends of one of your predicted genes</li></ul><ul id="caed3bff-2a2e-47aa-81b8-b6d437da8130" class="bulleted-list"><li>Match the start but not the end of a predicted gene</li></ul><ul id="35db43d8-6338-4410-8113-75da4059d87f" class="bulleted-list"><li>Match the end but not the start of a predicted gene</li></ul><ul id="b27abf04-9f78-4c9a-9c5c-5cc014d1eca5" class="bulleted-list"><li>Do not match neither the start not the end of a predicted gene</li></ul></li></ul><ul id="a42b74ef-1cc3-437b-8318-2fceaff428b7" class="bulleted-list"><li>The fraction of your predicted genes that:<ul id="283f92d8-aa2a-4bc3-b21b-28611e616e86" class="bulleted-list"><li>Perfectly match both ends of one of an annotated genes</li></ul><ul id="0575edc8-b312-419c-ab89-77ae8302a267" class="bulleted-list"><li>Match the start but not the end of an annotated gene</li></ul><ul id="0bf05541-5779-4df9-a956-be3173228804" class="bulleted-list"><li>Match the end but not the start of an annotated gene</li></ul><ul id="58809919-ab9b-42bd-8d27-25b82c6c80d5" class="bulleted-list"><li>Do not match neither the start not the end of an annotated gene</li></ul></li></ul><h3 id="d706a3ef-b97e-4976-8573-cdad81fd03b6" class="">Solution:</h3><p id="6cbce81b-3d7e-403a-8194-e54264c169f7" class="">Annotated genes that</p><ul id="f4acb4a1-616a-4f30-8945-b59d2ad8e480" class="bulleted-list"><li>Perfectly match both ends of a predicted gene: 0.86%</li></ul><ul id="54ffcc3a-9606-4f1f-9b29-5735c8e96ee3" class="bulleted-list"><li>Match the start but not the end of a predicted gene: 4.46%</li></ul><ul id="c89f1f14-deab-400d-a514-5794a3b5f687" class="bulleted-list"><li>Match the end but not the start of a predicted gene: 80.27%</li></ul><ul id="2d29f9bb-7006-463f-958d-65cde5ec852f" class="bulleted-list"><li>Don&#x27;t match neither the start nor end of a predicted gene: 14.41%</li></ul><p id="1c40124d-5ec2-4af8-8e05-23e5bdd04c39" class="">Predicted genes that</p><ul id="26109515-425a-4c80-97ba-eabc4d826648" class="bulleted-list"><li>Perfectly match both ends of an annotated gene: 0.59%</li></ul><ul id="bdd13784-c0ae-41e6-8166-0bb94f640194" class="bulleted-list"><li>Match the start but not the end of an annotated gene: 0.61%</li></ul><ul id="6d75195c-575d-4d93-a11a-db105bd803a6" class="bulleted-list"><li>Match the end but not the start of an annotated gene: 57.43%</li></ul><ul id="05cad48a-48ae-4d00-a993-38834f0f52cc" class="bulleted-list"><li>Don&#x27;t match neither the start nor end of an annotated gene: 41.36%</li></ul><p id="4ac02ac3-c24d-450d-9450-bdb204c25486" class="">
</p><p id="5ee734df-e77c-41c0-bebe-15dc77dbe9c0" class="">e) (15 points) What properties of annotated genes are associated to an elevated risk of being partially or completely missed by your predictor? What are the properties of genes predicted by your predictor that do not match an annotated gene? Write a paragraph for each question, supported by data and/or figures</p><h3 id="68ff3106-2c11-48a0-b693-cc90ad6c3d57" class="">Solution:</h3><p id="42ff500c-c226-4c76-b0ab-9bb3e36f90a1" class="">
</p><hr id="bc753167-9929-4245-8867-ff1e215164c1"/><h2 id="f6993bdd-0281-442f-9116-1ee5bca02ed3" class="">Question 2 (20 points).</h2><p id="e9ce2b72-7484-440a-966f-f306b120abe6" class="">If we think of an HMM as a machine to generate a random sequence of observations, the number of consecutive steps the path will remain at a given state follows a geometric distribution (https://en.wikipedia.org/wiki/Geometric_distribution). This means, for example, that, in our gene-finding HMM, the length distribution of exons, introns, and intergenic regions will be assumed to be geometric. However, in reality, these regions have length distributions that can be far from geometric. Consider the very simple two-state gene finding HMM shown below. Assume that we have a desired gene length distribution, provided in the form of a discrete probability distribution for lengths ranging from 1 to 1000:</p><figure id="8f8b9831-8847-4c98-b1ad-525833507023" class="equation"><style>@import url('https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.10.0/katex.min.css')</style><div class="equation-container"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math><semantics><mrow><mi>P</mi><mi>r</mi><mo>[</mo><mtext>length</mtext><mo>=</mo><mi>k</mi><mo>]</mo><mo>=</mo><msub><mi>p</mi><mi>k</mi></msub></mrow><annotation encoding="application/x-tex">Pr[\text{length}=k]=p_k</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathdefault" style="margin-right:0.13889em;">P</span><span class="mord mathdefault" style="margin-right:0.02778em;">r</span><span class="mopen">[</span><span class="mord text"><span class="mord">length</span></span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathdefault" style="margin-right:0.03148em;">k</span><span class="mclose">]</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:0.625em;vertical-align:-0.19444em;"></span><span class="mord"><span class="mord mathdefault">p</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.33610799999999996em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathdefault mtight" style="margin-right:0.03148em;">k</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span></span></span></span></span></div></figure><p id="c5d8190c-146b-47a8-9383-133b975d9c71" class="">Describe (in at most half a page) how you could modify the HMM below to produce a second HMM where the distribution over the duration of stay in the set of gene states is exactly the target length distribution. Note that the Gene state will probably need to be subdivided in several Gene sub-states. Describe not only the states of your HMM but also the transition probabilities.</p><p id="7a2ee454-a495-4513-bc72-cb83a56d6a3a" class="">
</p><figure id="8d400d13-0a7d-4511-bbf0-1dde1a9d9d1f" class="image"><a href="Assignment 3 Hidden Markov Models/Untitled.png"><img style="width:396px" src="Assignment 3 Hidden Markov Models/Untitled.png"/></a></figure><h3 id="b4c13176-23ef-41a6-bd55-0250d95372f8" class="">Solution:</h3><p id="b169a50b-a628-45cb-a3c7-c9b90560dd20" class="">From the above HMM, let <em>p</em> = probability of transitioning into a genic state.</p><p id="6ff5c8d4-c2dd-452f-86d9-a8257522eef7" class="">We want:</p><figure id="644cbd98-6262-42ad-bf5e-dbe181ed7231" class="equation"><style>@import url('https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.10.0/katex.min.css')</style><div class="equation-container"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math><semantics><mrow><mtext>product of probabilities up to k-th genic state</mtext><mo>=</mo><msub><mi>p</mi><mi>k</mi></msub><mspace linebreak="newline"></mspace><mo>⇒</mo><msup><mi>p</mi><mi>k</mi></msup><mo>=</mo><msub><mi>p</mi><mi>k</mi></msub><mspace linebreak="newline"></mspace><mo>⇒</mo><mi>p</mi><mo>=</mo><msubsup><mi>p</mi><mi>k</mi><mrow><mn>1</mn><mi mathvariant="normal">/</mi><mi>k</mi></mrow></msubsup></mrow><annotation encoding="application/x-tex">\text{product of probabilities up to k-th genic state}=p_k \\ \Rightarrow p^k=p_k \\ \Rightarrow p=p_k^{1/k}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.8888799999999999em;vertical-align:-0.19444em;"></span><span class="mord text"><span class="mord">product of probabilities up to k-th genic state</span></span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:0.625em;vertical-align:-0.19444em;"></span><span class="mord"><span class="mord mathdefault">p</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.33610799999999996em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathdefault mtight" style="margin-right:0.03148em;">k</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="mspace newline"></span><span class="base"><span class="strut" style="height:0.36687em;vertical-align:0em;"></span><span class="mrel">⇒</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:1.093548em;vertical-align:-0.19444em;"></span><span class="mord"><span class="mord mathdefault">p</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8991079999999999em;"><span style="top:-3.113em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathdefault mtight" style="margin-right:0.03148em;">k</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:0.625em;vertical-align:-0.19444em;"></span><span class="mord"><span class="mord mathdefault">p</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.33610799999999996em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathdefault mtight" style="margin-right:0.03148em;">k</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="mspace newline"></span><span class="base"><span class="strut" style="height:0.36687em;vertical-align:0em;"></span><span class="mrel">⇒</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:0.625em;vertical-align:-0.19444em;"></span><span class="mord mathdefault">p</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:1.3461079999999999em;vertical-align:-0.3013079999999999em;"></span><span class="mord"><span class="mord mathdefault">p</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.0448em;"><span style="top:-2.3986920000000005em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathdefault mtight" style="margin-right:0.03148em;">k</span></span></span><span style="top:-3.2198em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span><span class="mord mtight">/</span><span class="mord mathdefault mtight" style="margin-right:0.03148em;">k</span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.3013079999999999em;"><span></span></span></span></span></span></span></span></span></span></span></div></figure><p id="c346844f-cce7-4326-9996-e99919a981af" class="">However, the above is not feasible, because:</p><ol id="7defce5e-d363-445b-8c37-51fe51b286c5" class="numbered-list" start="1"><li><em>k</em> varies, while we only have 1 <em>p</em></li></ol><ol id="148f8dd9-4f2d-4db4-9bae-105ad1cfea95" class="numbered-list" start="2"><li>While generating a sequence, we have no way of knowing <em>k</em></li></ol><p id="4b4f8b43-1479-4c10-9279-8e6a1820bf63" class="">
</p><p id="a071b1b2-98e1-4c26-9dbc-4960a2e1d25f" class="">Therefore, we will break the 1 genic state into 1000 genic states in order to have 1000 genic transition probabilities.</p><p id="1017186d-0804-41cd-a4c1-90f6f4519e0d" class="">We define then:</p><figure id="80cf064e-c624-46b5-93aa-6e4acc5b8682" class="equation"><style>@import url('https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.10.0/katex.min.css')</style><div class="equation-container"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math><semantics><mrow><msub><mi>G</mi><mi>k</mi></msub><mtext> genic state at length k</mtext><mspace linebreak="newline"></mspace><mover accent="true"><mi>G</mi><mo stretchy="true">‾</mo></mover><mtext> non-genic state</mtext><mspace linebreak="newline"></mspace><mi>T</mi><mo>(</mo><mi>i</mi><mo>)</mo><mtext> probability of transitioning into genic state at length </mtext><mi>i</mi></mrow><annotation encoding="application/x-tex">G_k\text{ genic state at length k}\\\overline{G}\text{ non-genic state}\\T(i)\text{ probability of transitioning into genic state at length }i</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.8888799999999999em;vertical-align:-0.19444em;"></span><span class="mord"><span class="mord mathdefault">G</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.33610799999999996em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathdefault mtight" style="margin-right:0.03148em;">k</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord text"><span class="mord"> genic state at length k</span></span></span><span class="mspace newline"></span><span class="base"><span class="strut" style="height:1.0777700000000001em;vertical-align:-0.19444em;"></span><span class="mord overline"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8833300000000001em;"><span style="top:-3em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord mathdefault">G</span></span></span><span style="top:-3.80333em;"><span class="pstrut" style="height:3em;"></span><span class="overline-line" style="border-bottom-width:0.04em;"></span></span></span></span></span></span><span class="mord text"><span class="mord"> non-genic state</span></span></span><span class="mspace newline"></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathdefault" style="margin-right:0.13889em;">T</span><span class="mopen">(</span><span class="mord mathdefault">i</span><span class="mclose">)</span><span class="mord text"><span class="mord"> probability of transitioning into genic state at length </span></span><span class="mord mathdefault">i</span></span></span></span></span></div></figure><p id="11356132-41c1-4d9c-a6c9-e08bc6fbbea6" class="">Then, for a genic region of length k, the probability is:</p><figure id="d19a1927-82a1-4eaf-89f2-d00a576a3344" class="equation"><style>@import url('https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.10.0/katex.min.css')</style><div class="equation-container"><span class="katex-display"><span class="katex"><span class="katex-mathml"><math><semantics><mrow><munderover><mo>∏</mo><mi>i</mi><mi>k</mi></munderover><mi>T</mi><mo>(</mo><mi>i</mi><mo>)</mo><mo>=</mo><msub><mi>p</mi><mi>k</mi></msub></mrow><annotation encoding="application/x-tex">\prod_i^k T(i)=p_k</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:3.1137820000000005em;vertical-align:-1.277669em;"></span><span class="mop op-limits"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.8361130000000003em;"><span style="top:-1.872331em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathdefault mtight">i</span></span></span><span style="top:-3.050005em;"><span class="pstrut" style="height:3.05em;"></span><span><span class="mop op-symbol large-op">∏</span></span></span><span style="top:-4.3000050000000005em;margin-left:0em;"><span class="pstrut" style="height:3.05em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathdefault mtight" style="margin-right:0.03148em;">k</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:1.277669em;"><span></span></span></span></span></span><span class="mspace" style="margin-right:0.16666666666666666em;"></span><span class="mord mathdefault" style="margin-right:0.13889em;">T</span><span class="mopen">(</span><span class="mord mathdefault">i</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2777777777777778em;"></span></span><span class="base"><span class="strut" style="height:0.625em;vertical-align:-0.19444em;"></span><span class="mord"><span class="mord mathdefault">p</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.33610799999999996em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathdefault mtight" style="margin-right:0.03148em;">k</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span></span></span></span></span></div></figure><p id="4d8891b7-77c7-45dc-abf3-50c0ee08a414" class="">The above is accomplished through the following HMM:</p><p id="5b5ec5db-3fd5-4534-b798-6e07f73bb985" class="">
</p><figure id="fd6952f9-5394-4764-a290-e0746e975bb0" class="image"><a href="Assignment 3 Hidden Markov Models/Untitled 1.png"><img style="width:576px" src="Assignment 3 Hidden Markov Models/Untitled 1.png"/></a></figure><p id="cec66577-12a9-482e-a905-5758d8b1bc0b" class="">The transition probability matrix is:</p><figure id="2f9a1dd4-9408-4df7-9d58-207fdca61382" class="image"><a href="Assignment 3 Hidden Markov Models/Untitled 2.png"><img style="width:576px" src="Assignment 3 Hidden Markov Models/Untitled 2.png"/></a></figure></div></article></body></html>